ZCS-6871 Make OWASP filtering configurable

chinmay15 opened the pull request:
**Problem:** Make OWASP filtering configurable

**Fix:**
- Created OwaspHtmlSanitizer class for owasp filtering which will be implemented later
- Created zimbraUseOwaspHtmlSanitizer, boolean ldap attribute at globalConfig level
- Updated DefangFactory class to use zimbraUseOwaspHtmlSanitizer to decide which defang class to use for filtering.

**Testing done:**
- Build successful.
- zimbraUseOwaspHtmlSanitizer is available at global config level with default value as TRUE
- Added unit test
[report.zip](https://github.com/Zimbra/zm-mailbox/files/3062667/report.zip)


**Testing to be done by QA:**
- Make sure zimbraUseOwaspHtmlSanitizer is available at global config level with default value TRUE.
- If zimbraUseOwaspHtmlSanitizer set false, HtmlDefang filter should be used.

rupalid commented:
Unit test?

shrikant-prasad commented on store/src/java/com/zimbra/cs/service/mail/ToXML.java:
```diff
@@ -2295,7 +2295,8 @@ public static Element encodeInviteComponent(Element parent, ItemIdFormatter ifmt
                 }
                 String descHtml = invite.getDescriptionHtml();
-                BrowserDefang defanger = DefangFactory.getDefanger(MimeConstants.CT_TEXT_HTML);
+                boolean isOwasp = Provisioning.getInstance().getConfig().getBooleanAttr(Provisioning.A_zimbraUseOwaspHtmlSanitizer, Boolean.TRUE);
```
Please enhance the DefangFactory.getDefanger() to check the configuration and accordingly return the defanger.

chinmay15 replied:
done

chinmay15 commented:
merged in feature/owasp branch 

commit 81106295c3c00b9adaf1ec919a5ffbbaf7ebd9a4
Author: Chinmay Pathak <15.chinmay@gmail.com>
Date:   Tue Apr 9 14:24:16 2019 +0530

    ZCS-6871 Make OWASP filtering configurable
    
    Added unit test
    
    updated comment as per review comment
    
    Resolving review comments
